src/app/qgslabelinggui.cpp

```diff
@@ -668,3 +654,128 @@ void QgsLabelingGui::deactivateField( QgsPalLayerSettings::Property key )
     mDataDefinedProperties.setProperty( key, p );
   }
 }
+
+void QgsLabelingGui::updateGeometryTypeBasedWidgets()
+{
+  QgsWkbTypes::GeometryType geometryType;
+
+  if ( mGeometryGeneratorGroupBox->isChecked() )
+    geometryType = mGeometryGeneratorType->currentData().value<QgsWkbTypes::GeometryType>();
+  else
+    geometryType = mLayer->geometryType();
+
+  // show/hide options based upon geometry type
+  chkMergeLines->setVisible( geometryType == QgsWkbTypes::LineGeometry );
+  mDirectSymbolsFrame->setVisible( geometryType == QgsWkbTypes::LineGeometry );
+  mMinSizeFrame->setVisible( geometryType != QgsWkbTypes::PointGeometry );
+  mPolygonObstacleTypeFrame->setVisible( geometryType == QgsWkbTypes::PolygonGeometry );
+  mPolygonFeatureOptionsFrame->setVisible( geometryType == QgsWkbTypes::PolygonGeometry );
+
+
+  // set placement methods page based on geometry type
+  switch ( geometryType )
+  {
+    case QgsWkbTypes::PointGeometry:
+      stackedPlacement->setCurrentWidget( pagePoint );
+      break;
+    case QgsWkbTypes::LineGeometry:
+      stackedPlacement->setCurrentWidget( pageLine );
+      break;
+    case QgsWkbTypes::PolygonGeometry:
+      stackedPlacement->setCurrentWidget( pagePolygon );
+      break;
+    case QgsWkbTypes::NullGeometry:
+      break;
+    case QgsWkbTypes::UnknownGeometry:
+      qFatal( "unknown geometry type unexpected" );
+  }
+
+  if ( geometryType == QgsWkbTypes::PointGeometry )
+  {
+    // follow placement alignment is only valid for point layers
+    if ( mFontMultiLineAlignComboBox->findText( tr( "Follow label placement" ) ) == -1 )
+      mFontMultiLineAlignComboBox->addItem( tr( "Follow label placement" ) );
+  }
+  else
+  {
+    int idx = mFontMultiLineAlignComboBox->findText( tr( "Follow label placement" ) );
+    if ( idx >= 0 )
+      mFontMultiLineAlignComboBox->removeItem( idx );
+  }
+}
+
+void QgsLabelingGui::showGeometryGeneratorExpressionBuilder()
+{
+  QgsExpressionBuilderDialog expressionBuilder( mLayer );
+
+  expressionBuilder.setExpressionText( mGeometryGenerator->text() );
+  expressionBuilder.setExpressionContext( createExpressionContext() );
+
+  QgsDistanceArea da;
+  da.setSourceCrs( mLayer->crs(), QgsProject::instance()->transformContext() );
+  da.setEllipsoid( QgsProject::instance()->ellipsoid() );
+  expressionBuilder.setGeomCalculator( da );
```

Comment:
This is no longer required -- if your expression context correctly contains a project + layer scope, it will be handled for you.